File: frontend/src/lib/scamshield/repair-urls.js
// URL-repair pass: reconstruct URLs mangled by OCR (dropped/spaced punctuation)
// so the strict https?:// extractor inside pipeline.js can recognize them.
// Applied to OCR-extracted text BEFORE it enters the pipeline, on both JS and
// Python OCR paths, so OCR-engine differences don't silence URL phishing signals.

const SEG = /[a-z0-9][a-z0-9-]*/i;
// "http"/"htp"(+ typo variants hitp/htip) rarely appear inside a normal word,
// so no strict word boundary: lets OCR space-merges ("athttp"), letter-drops
// ("htp") and transpositions ("hitp","htip") still be seen as URL starts.
const ANCHOR_RE = /(?:https?|htps?|htp|hitp|htip)[\s:/.\\|_’‘]{0,4}/gi;

// Known/generic TLDs. A whitespace-separated trailing token may rejoin the host
// as a dotted TLD ONLY if it is in this set (e.g. OCR rendering "host.in" as
// "host in"). This avoids swallowing sentence prose that follows a URL.
const TLD_SET = new Set([
  "xyz", "in", "tk", "com", "co", "net", "org", "info", "io", "me", "tv",
  "us", "uk", "ca", "au", "de", "fr", "es", "it", "mx", "br", "ru", "top",
  "online", "site", "app", "store", "cloud", "live", "world", "club", "asia",
  "gov",
]);

const MAX_HOST_SEGMENTS = 6;
const NON_SPACE_SEP = /[.:/\\|’‘]/;

function isSegStart(c) {
  return !!c && /[a-z0-9]/i.test(c);
}

// Read one segment ([a-z0-9][a-z0-9-]*) starting at s[pos]; returns
// {seg, next} where next is the index just past the segment.
function readSeg(s, pos) {
  const m = SEG.exec(s.slice(pos));
  if (!m || m.index !== 0) return null;
  return { seg: m[0], next: pos + m[0].length };
}

function tryParseUrl(s, start) {
  const head = s.slice(start, start + 8).toLowerCase();
  let scheme;
  let j;
  if (head.startsWith("https")) {
    scheme = "https";
    j = start + 5;
  } else if (head.startsWith("http")) {
    scheme = "http";
    j = start + 4;
  } else if (head.startsWith("hitp") || head.startsWith("htip") || head.startsWith("htp")) {
    // OCR typos: hitp/htip (transposition) or htp (letter-drop)
    scheme = head.startsWith("htps") ? "https" : "http";
    j = start + 3;
    if (head.startsWith("hitp") || head.startsWith("htip")) j = start + 4;
  } else if (head.startsWith("www")) {
    scheme = "www";
    j = start + 3;
  } else {
    return null;
  }

  const n = s.length;
  // skip partially-eaten "://" and stray smart-quote artifacts
  let rp = s.slice(j, j + 16).match(/^[\s:/.\\|_’‘]*/);
  j += rp ? rp[0].length : 0;

  const segs = [];
  let k = j;

  // Always take the first host segment.
  const first = readSeg(s, k);
  if (!first) return null;
  segs.push(first.seg);
  k = first.next;

  while (segs.length < MAX_HOST_SEGMENTS) {
    // Case A: non-space separator(s) already consumed right after a segment,
    // meaning a real dotted/path continuation (http://a.b /c /d).
    // Case B: whitespace gap — possible missing dot before a TLD.
    let scroll = 0;
    while (k + scroll < n && NON_SPACE_SEP.test(s[k + scroll])) scroll++;

    if (scroll > 0 && k + scroll < n && isSegStart(s[k + scroll])) {
      // dot/slash-joined continuation
      k += scroll;
      const nxt = readSeg(s, k);
      segs.push(nxt.seg);
      k = nxt.next;
      continue;
    }

    // Not a dotted continuation. Check whitespace-separated TLD rejoin.
    if (k < n && /\s/.test(s[k])) {
      let w = k;
      while (w < n && /\s/.test(s[w])) w++;
      while (w < n && NON_SPACE_SEP.test(s[w])) w++; // stray '.' before TLD
      const nxt = readSeg(s, w);
      if (nxt && TLD_SET.has(nxt.seg.toLowerCase()) && segs.length >= 1) {
        segs.push(nxt.seg);
        k = nxt.next; // stop right after the TLD; do NOT consume following text
      }
    }
    break;
  }

  const host = segs.join(".");
  const url = scheme === "www" ? "www." + host : scheme + "://" + host;
  return { start, end: k, url };
}

function repairUrls(text) {
  if (!text || typeof text !== "string") return text;
  const replacements = [];
  let m;
  ANCHOR_RE.lastIndex = 0;
  while ((m = ANCHOR_RE.exec(text)) !== null) {
    const parsed = tryParseUrl(text, m.index);
    if (parsed && parsed.url && parsed.end > parsed.start) {
      replacements.push(parsed);
      ANCHOR_RE.lastIndex = Math.max(ANCHOR_RE.lastIndex, parsed.end);
    }
  }
  replacements.sort((a, b) => b.start - a.start);
  let out = text;
  for (const r of replacements) {
    // If the URL was merged into a preceding word (e.g. OCR turned "at http"
    // into "athttp"), restore the lost word boundary so downstream tokenization
    // and the strict extractor see a clean URL token.
    let prefix = "";
    if (r.start > 0 && /[\w]/.test(out[r.start - 1])) prefix = " ";
    out = out.slice(0, r.start) + prefix + r.url + out.slice(r.end);
  }
  return out;
}

export { repairUrls };

File: experiments/repair_urls.py
"""URL-repair pass mirroring repair_urls.js for the Python OCR path.

Reconstructs URLs mangled by OCR (dropped/spaced punctuation) so the strict
https?:// extractor can recognize them. Applied to OCR-extracted text BEFORE it
enters the pipeline so both OCR engines are treated consistently.
"""

import re

_SEG = re.compile(r"[a-z0-9][a-z0-9-]*", re.I)
# "http"/"htp"(+ typos hitp/htip) rarely appear inside a normal word, so no
# strict word boundary: catches OCR space-merges, letter-drops and transposes.
_ANCHOR_RE = re.compile(r"(?:https?|htps?|htp|hitp|htip)[\s:./\\|_\u2018\u2019]{0,4}", re.I)

# Known/generic TLDs (mirrors repair_urls.js).
_TLD_SET = frozenset({
    "xyz", "in", "tk", "com", "co", "net", "org", "info", "io", "me", "tv",
    "us", "uk", "ca", "au", "de", "fr", "es", "it", "mx", "br", "ru", "top",
    "online", "site", "app", "store", "cloud", "live", "world", "club", "asia",
    "gov",
})

_MAX_HOST_SEGMENTS = 6
_NON_SPACE_SEP = re.compile(r"[.:/\\|\u2018\u2019]")


def _is_seg_start(c):
    return bool(c and (c.isalnum()))


def _read_seg(s, pos):
    m = _SEG.match(s, pos)
    if not m:
        return None
    return {"seg": m.group(0), "next": m.end()}


def _try_parse_url(s, start):
    head = s[start:start + 8].lower()
    if head.startswith("https"):
        scheme, j = "https", start + 5
    elif head.startswith("http"):
        scheme, j = "http", start + 4
    elif head.startswith("hitp") or head.startswith("htip") or head.startswith("htp"):
        scheme = "https" if head.startswith("htps") else "http"
        j = start + 3
        if head.startswith("hitp") or head.startswith("htip"):
            j = start + 4
    elif head.startswith("www"):
        scheme, j = "www", start + 3
    else:
        return None

    n = len(s)
    rp = re.match(r"^[\s:./\\|_\u2018\u2019]*", s[j:j + 16])
    j += len(rp.group(0)) if rp else 0

    segs = []
    k = j

    first = _read_seg(s, k)
    if not first:
        return None
    segs.append(first["seg"])
    k = first["next"]

    while len(segs) < _MAX_HOST_SEGMENTS:
        scroll = 0
        while k + scroll < n and _NON_SPACE_SEP.match(s[k + scroll]):
            scroll += 1

        if scroll > 0 and k + scroll < n and _is_seg_start(s[k + scroll]):
            k += scroll
            nxt = _read_seg(s, k)
            segs.append(nxt["seg"])
            k = nxt["next"]
            continue

        if k < n and s[k].isspace():
            w = k
            while w < n and s[w].isspace():
                w += 1
            while w < n and _NON_SPACE_SEP.match(s[w]):
                w += 1
            nxt = _read_seg(s, w)
            if nxt and nxt["seg"].lower() in _TLD_SET and len(segs) >= 1:
                segs.append(nxt["seg"])
                k = nxt["next"]
        break

    host = ".".join(segs)
    url = ("www." + host) if scheme == "www" else (scheme + "://" + host)
    return {"start": start, "end": k, "url": url}


def repair_urls(text):
    if not text or not isinstance(text, str):
        return text
    replacements = []
    for m in _ANCHOR_RE.finditer(text):
        parsed = _try_parse_url(text, m.start())
        if parsed and parsed["url"] and parsed["end"] > parsed["start"]:
            replacements.append(parsed)
    replacements.sort(key=lambda r: r["start"], reverse=True)
    out = text
    for r in replacements:
        prefix = ""
        if r["start"] > 0 and out[r["start"] - 1].isalnum():
            prefix = " "
        out = out[:r["start"]] + prefix + r["url"] + out[r["end"]:]
    return out
